경북대 BE_배준호 2주차 과제 (1단계)

BaeJunH0 opened the pull request:
step 1 요구 사항과 같이 제약 조건을 추가하였습니다.
이를 위해 POST 요청에 @RequestBody을 적용하고, ajax 송수신 데이터 타입을 application/json 으로 설정하였습니다.
error 객체가 전송되는 것을 읽기 위해 responseText와 JSON.parse를 통해 js에서 객체를 읽고 사용할 수 있게 하였습니다.
JS 연결은 다양한 방법 ( ./부터 시작하는 상대경로, 소스루트부터 시작하는 경로 적용 ) 을 적용해보았는데 잘 안되었습니다.

Commit:
feat : step 0. 기본 코드 준비

피드백 받은 부분을 반영한 코드로 프로젝트 초기화

Commit:
refactor : 상품 수정 기능 AJAX로 변경

View의 형식을 변경하여 상품 수정 기능을 AJAX 방식으로 요청하는 것으로 변경

Commit:
feat : Implement Validation

상품 이름에 대한 유효성 검사를 구현하였다

Commit:
chore : 주석 추가

Commit:
feat : Update 기능에서도 Validation 추가

Update에서도 product의 name 필드를 사용하니, 여기에도 유효성 검사 부분을 넣아주었다

BaeJunH0 committed:
Merge branch 'baejunh0' into step1

Commit:
Merge branch 'step1' of https://github.com/BaeJunH0/spring-gift-wishlist into step1

BaeJunH0 commented:
추가적으로 PUT 요청에서도 Product 클래스의 name 필드 ( 유효성 검증 대상 ) 이 사용되어서 해당 부분도 유효성 검사 부분을 추가했습니다.
또, @Valid 어노테이션에 의해 발생된 BindingResult 클래스를 ControllerAdvice에서 ExceptionHandler를 사용하여 유효성 검사에서 발생한 예외 사항을 한 번에 처리하도록 하여 코드의 중복을 줄였습니다

lsm7179 commented on src/main/java/gift/DTO/Product.java:
```diff
@@ -1,7 +1,16 @@
+
 public class Product {
     private Long id;
+    @Size(min=1, max=15, message = "상품명은 1자 이상 15자 이하여야 합니다")
+    @Pattern.List({
+        @Pattern(regexp = "^[가-힣a-zA-Z0-9\\s\\(\\)\\[\\]\\+\\-\\&\\/\\_]+$",
+                message = "특수문자는 ( ), [ ], +, -, &, /, _ 만 허용되며, 한글, 영어, 숫자만 입력 가능합니다."),
+        @Pattern(regexp = "^(?!.*카카오).*$", message = "'카카오'가 포함된 상품명은 MD와 협의 후 사용해주시길 바랍니다")
```
@Pattern.List와 @Pattern을 잘 활용해주었네요~!

BaeJunH0 replied:
감사합니다 ㅎㅎ

lsm7179 commented on src/main/java/gift/controller/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@ControllerAdvice
```
RestControllerAdvice과 차이를 알고 계실까요?
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/RestControllerAdvice.html

BaeJunH0 replied:
> [https://medium.com/sjk5766/restcontrolleradvice-vs-controlleradvice-정리-1591010ce0b0](https://medium.com/sjk5766/restcontrolleradvice-vs-controlleradvice-%EC%A0%95%EB%A6%AC-1591010ce0b0)
> 
- RestControllerAdvice와 ControllerAdvice의 차이
    - RestController와 Conroller의 차이와 동일합니다
        - 반환 값이 data인지 view인지 ( **View Resolver에게 전달될지 아닐지** )
    - 기존 코드에서 문제가 생기지 않은 것은, 아마도 RestController에서 발생한 Exception 처리라서 Spring이 잘 처리해준 것으로 예상됩니다

lsm7179 replied:
네 좋습니다. 
왜 굳이 스프링에서는 RestControllerAdvice와 ControllerAdvice 각각 제공할까요?
개발자는 필요하다면 ControllerAdvice와 Conroller어노테이션을 상속받아서 커스텀 어노테이션으로 만들어서 쓰면 됩니다.

그런데 굳이 스프링은 rest 어노테이션을 별도로 만들어 줬어요. 왜일까요?
한번 고민해보세요~

lsm7179 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -38,9 +43,10 @@ public ResponseEntity<List<Product>> findProducts() {
      */
     @PostMapping("/api/products")
-    public ResponseEntity<Void> addProduct(Product product){
+    public ResponseEntity<Map<String, String>> addProduct(@Valid @RequestBody Product product){
```
Product가 도메인객체와 dto역할 두가지를 하는거 같아요.😨
단일책임원칙으로 response, request 객체를 별도로 사용해서 validate를 도메인 객체로 부터 분리해 보세요.😃

lsm7179 commented on src/main/resources/templates/index.html:
```diff
@@ -4,7 +4,7 @@
     <title>상품 목록</title>
     <link rel="stylesheet" href="/style/index_style.css">
-    <script type="text/javascript" src="http://code.jquery.com/jquery-3.5.1.min.js"></script>
+    <script type="text/javascript" src="http://code.jquery.com/jquery-3.6.0.min.js"></script>
```
웹 백엔드 교육이니 js쪽이 힘들다면 당장 분리하지 않아도 될거 같아요.🤗
그래도 꼭 분리해보고 싶으시면 DM주세요~

BaeJunH0 replied:
넵! 사실 엄~청 중요한 부분은 아니니까 우선순위 반영해서 백엔드에 더 집중하겠습니다!

lsm7179 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -38,9 +43,10 @@ public ResponseEntity<List<Product>> findProducts() {
      */
     @PostMapping("/api/products")
-    public ResponseEntity<Void> addProduct(Product product){
+    public ResponseEntity<Map<String, String>> addProduct(@Valid @RequestBody Product product){
```
Map<String, String>으로 받아서 처리할 이유가 있나요?

BaeJunH0 replied:
이 부분은 ControllerAdvice를 만들어서 Error로 인한 Exception을 따로 다루기 전에, 해당 메서드에서 BindingResult 클래스를 이용해서 Error을 처리하던 흔적입니다, 지우는 것을 깜빡했네요! 수정하겠습니다!

lsm7179 reviewed (changes requested):
준호님.
1단계 요구사항대로 잘 이행되었어요~!👍
간단한 피드백을 남겼으니 피드백을 해결 해보면 좋을거 같습니다.
금방하실수 있을거에요.

Commit:
chore : Service 클래스 설명 추가

Commit:
refactor : 피드백 내용 반영 + @

DTO 클래스와 Domain model (Entity) 클래스를 분리, 오늘 배운 내용을 적용 하여 Repository-Service-Controller 구조 완성

lsm7179 commented on src/main/java/gift/DTO/Product.java:
```diff
@@ -1,7 +1,17 @@
+import jakarta.validation.constraints.Pattern;
+import jakarta.validation.constraints.Size;
+
 public class Product {
```
해당 클래스명이 도메인처럼 느껴지네요.
dto라면 특히 request라면 ProductRequest 는 어떤가요?

lsm7179 commented on src/main/java/gift/domain/ProductEntity.java:
```diff
@@ -0,0 +1,56 @@
+
+import gift.DTO.Product;
+
+public class ProductEntity {
```
해당 클래스명은 Product가 더 적절해보이네요.😀

lsm7179 commented on src/main/java/gift/domain/ProductEntity.java:
```diff
@@ -0,0 +1,56 @@
+        this.imageUrl = imageUrl;
+    }
+
+    public ProductEntity(Product productDTO) {
+        this.id = productDTO.getId();
+        this.name = productDTO.getName();
+        this.price = productDTO.getPrice();
+        this.imageUrl = productDTO.getImageUrl();
+    }
```
dto가 도메인을 침범했네요.

```
public ProductEntity(Long id, String name, int price, String imageUrl) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.imageUrl = imageUrl;
    }
```
이 생성자로도 충분히 객체생성이 가능하게 수정해보아요.

lsm7179 commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -1,54 +1,64 @@
-    private final JdbcTemplate jdbcTemplate;
-    public ProductService(JdbcTemplate jdbcTemplate){
-        this.jdbcTemplate = jdbcTemplate;
+    private final ProductRepository productRepository;
+    public ProductService(ProductRepository productRepository){
```
코드 컨벤션을 지켜보세요~
각 멤버 변수 생성자 사이에 공백 라인을 넣어보세요~

lsm7179 reviewed (approved):
준호님.
1단계 고생하셨어요.👍
머지 할 수 있게 요구사항을 잘 이행 해주신거 같아요.🔥

간단한 피드백이랑 고민거리도 넣었습니다. 다음 단계에서 같이 반영해보세요~!👍
궁금한게 있으면 언제든 편하게 DM주세요~
